### ISSUE TITLE
Issues with 0.10.1 release's new fork behavior

### ISSUE THREAD

**Issue by joonhyublee** [yelouafi/redux-saga#291]
Hi,
I was just trying the brand new `0.10.1` release, and I'm sorry to report that basically my whole app ended up broken. I have multiple points of failures throughout my app, so it's impossible to say for now which parts are broken, or whether it's something wrong with the new release, or just the wrong way i've been using saga, but I have suspicion that it's the new `fork` behavior from https://github.com/yelouafi/redux-saga/issues/277 that caused this.

Let me explain. This is what I had, which worked in `0.10.0`:

```javascript
export default function* () {
    const chains_dictionary = {};
    
    while (true) {
        
        const action = take('ACTION');
        const { chain_name } = action;
            
        // if there is already a chain, append to that chain
        if (chains_dictionary[chain_name]) {
            chains_dictionary[chain_name] = yield fork(chain, chains_dictionary[chain_name], call(each_chain_saga, action));
        }
        
        // if there isn't already a chain, start a chain with a no-op task
        else {
            console.log('starting a chain');
            chains_dictionary[chain_name] = yield fork(chain, null, call(each_chain_saga, action));
        }
        
        console.log('loop returning');
    }
    
}
```
where chain and no_op are:
```javascript
export function* no_op () {}

// new task = (perform old task --> and then perform next task)
export function* chain (task, effect) {
    if (task) yield join(task);
    yield effect;
}
```
Before with `0.10.0` it used to be:
```
ACTION
"starting a chain"
"loop returning"
ACTION
"starting a chain"
"loop returning"
ACTION
"starting a chain"
"loop returning"
ACTION
"starting a chain"
"loop returning"
```
but now with `0.10.1` it is:
```
ACTION
ACTION --> debounced
ACTION --> debounced
ACTION --> debounced
"starting a chain"
"loop returning"
```
In other words,  in `0.10.1` the `while` loop is not looping fast enough to take all the rapidly firing `ACTION`s, even though it's unblocking, debouncing `ACTION`s as a result, whereas it used to be fast enough before in `0.10.0`. Is this expected?

Added: I got things working again with the new `actionChannel` catching and queuing everything, so are we saying now that **only `actionChannel` can guarantee no debouncing?**

**Comment by yelouafi** [comment]
This is indeed related to the way #277 was fixed. Prior  to 0.10.1 forks run its body before the parent resume causing #277. In 0.10.1 the fork is enqueued until the parent resumes first. In your example the parent resumes and takes the next action before even the fork starts. I guess because actions are dispatched synchronously. So it's because is 'too fast' and not 'not fast enough' (unless I'm mistaken)

So we have to choose which way to go. Because it seems we can't handle #277 and this at the same time.

Could you tell more about the use case here: why  do you need to debounce actions fired synchronously? Typically debouncing is applied to UI events which always are handled in different event loops?

**Comment by yelouafi** [comment]
Sorry. Just forget about the previous comment

**Comment by joonhyublee** [comment]
@yelouafi Okay, I'll disregard your first comment, which I couldn't understand because if parent was resuming fast, it should still be logging `"starting a chain"` and `"loop returning"`..? Below is my use case, in which it is crucial that all actions are caught and no action is debounced:
```
             uninterruptable chained task
               (joined the first time)     ┌──( repeated many times )───┐       (left the last time)
              ┌────────────────────────┐     ┌────────┐     ┌─────────┐     ┌──────────────────────────┐
users saga ─┬─┤ join --> join complete │-->..│ join+1 │-->..│ leave-1 │-->..│ leave --> leave complete │
     	    │ └────────────────────────┘     └────────┘     └─────────┘     └──────────────────────────┘
   forked   ├─ ...
  for each  │
    user    ├─ ...
            │
            :
            :
```
To give you some context, I have multiple react components freely subscribing/unsubscribing to real-time user data backend built with websocket. Because subscribing to the same user more than once is redundant, I keep count of how many components are listening to the same user, so that I can start a new subscription when there is 1 listener, do nothing when there is 2, 3, 4, ... listeners, and unsubscribe when there is 0 listener again. For app-wide consistency, that count is stored in the redux store, with `put` effects from individual `user saga` forked and chained by `users saga`(plural).

As you can see, for resource management it is crucial that the count is accurate, and no join/leave action should go unhandled. **So I need some guarantee all such actions are caught,** which I think, the only way for now is using `actionChannel`. However, as I mentioned in https://github.com/yelouafi/redux-saga/issues/277, I think `takeEvery` can be a misnomer when, in certain cases, actions can go missing.

I'll more thoroughly read https://github.com/yelouafi/redux-saga/issues/235 and the source code, to see if I can come up with a suggestion for this.

**Comment by yelouafi** [comment]
@joonhyublee can you try with the new 0.10.2? I don't have enough info about your issue but as I said in #277 there was an issue with the impl of the fix in 0.10.1

**Comment by joonhyublee** [comment]
@yelouafi I can confirm that 0.10.2 now works with my use case too. Whatever you did, it worked! Thanks!!
